### PR TITLE
Enhancing the preview-middleware to support edit modes 

### PR DESCRIPTION
**Pull request by tobiasqueck**
## Description
Please note this is a breaking change because the possibility to change the generated html based on url parameter to use one endpoint for preview and editing has been removed. Instead, from now on, only the yaml configuration affects the generated templates and url parameters affect the runtime behavior.

## Changes
* `preview-middleware-client`
  * added placeholder for `adp` (adaptation projects) and `cpe` (control property editor) 
  * improved the jest test setup
*  `preview-middleware`
  * remove modification of the generated html based on url parameter
  * enhance config with an `rta` object allowing to define mount points for editing 
* `adp-tooling`
  * updated the generator template to use the new `preview-middleware` version and configuration

### CODE REVIEW

File: packages/preview-middleware-client/src/cpe/init.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import log from 'sap/base/Log';
+import type RuntimeAuthoring from 'sap/ui/rta/RuntimeAuthoring';
+
+export default function(rta: RuntimeAuthoring) {
```
Comment (zdravko-georgiev) [inline packages/preview-middleware-client/src/cpe/init.ts:4]: Don't we need `init(rta)` also here?

Comment (tobiasqueck) [reply packages/preview-middleware-client/src/cpe/init.ts:4]: Nope. `flp/init` is called as part of the initialization of the FLP sandbox. If RTA is enabled and a plugin script is configured then it is called by RTA (not by us) with RTA as parameter. So `cpe/init` is called by RTA already.

File: packages/preview-middleware/src/base/flp.ts [hunk 2]
```diff
@@ -145,13 +148,51 @@ export class FlpSandbox {
...
+                        previewUrl: `${previewUrl}?sap-ui-xx-viewCache=false&fiori-tools-rta-mode=forAdaptation&sap-ui-rta-skip-flex-validation=true&sap-ui-xx-condense-changes=true#${this.config.intent.object}-${this.config.intent.action}`
+                    });
+                    res.status(200).contentType('html').send(html);
+                }) as RequestHandler);
+            }
+            this.router.get(previewUrl, (async (_req: Request, res: Response) => {
+                const config = { ...this.templateConfig };
+                config.flex = {
+                    layer: rta.layer,
+                    ...rta.options,
```
Comment (voicis) [inline packages/preview-middleware/src/base/flp.ts:185]: Is passing all options from the yaml file to RTA intentional?

File: packages/preview-middleware/test/unit/base/__snapshots__/flp.test.ts.snap [hunk 3]
```diff
@@ -258,7 +259,39 @@ exports[`FlpSandbox router test/flp.html 1`] = `
...
+            }
+            iframe {
+                width: 100%;
+                height: 100%;
+                border: none;
+            }
+        </style>
+    </head>
+    <body>
+        <iframe src=\\"/my/editor.html.inner.html?sap-ui-xx-viewCache=false&fiori-tools-rta-mode=forAdaptation&sap-ui-rta-skip-flex-validation=true&sap-ui-xx-condense-changes=true#app-preview\\"></iframe>
```
Comment (voicis) [inline packages/preview-middleware/test/unit/base/__snapshots__/flp.test.ts.snap:288]: Having the `html` extension in the middle of name `editor.html.inner.html` looks a bit odd.

File: packages/preview-middleware-client/src/cpe/init.ts (hunk 1 above)
Comment (zdravko-georgiev) [reply packages/preview-middleware-client/src/cpe/init.ts:4]: OK 👍

File: packages/preview-middleware-client/test/__mock__/window.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        getCore: jest.fn().mockReturnValue(sapCoreMock),
+        require: {
+            toUrl: jest.fn()
+        },
+        loader: {
+            config: jest.fn()
+        }
+    },
+    ushell: {
+        Container: {
```
[the remark this answers is not in the document]
Comment (tobiasqueck) [reply packages/preview-middleware-client/test/__mock__/window.ts:20]: Sadly no, that is actually a globally initialized variable.

File: packages/preview-middleware-client/types/sap.ui.rta.d.ts [hunk 5]
```diff
@@ -98,15 +90,38 @@ declare module 'sap/ui/rta/RuntimeAuthoring' {
...
+
+   interface RuntimeAuthoring {
+        attachSelectionChange(handler: (event: SelectionChangeEvent) => void): void;
         attachModeChanged: (handler: (event: Event) => void) => void;
         attachUndoRedoStackModified: (handler: (event: Event) => void) => void;
         getCommandStack: () => Stack;
         getService: <T>(name: 'outline' | string) => Promise<T>;
         getSelection: () => ElementOverlay[];
         getDefaultPlugins: () => { contextMenu: ContextMenu };
+        getFlexSettings: () => {
```
[the remark this answers is not in the document]
Comment (tobiasqueck) [reply packages/preview-middleware-client/types/sap.ui.rta.d.ts]: I missed that one, fixed with c3a8da0

File: packages/preview-middleware-client/types/sap.ui.rta.d.ts [hunk 6]
```diff
@@ -98,15 +90,38 @@ declare module 'sap/ui/rta/RuntimeAuthoring' {
...
+   interface RuntimeAuthoring {
+        attachSelectionChange(handler: (event: SelectionChangeEvent) => void): void;
         attachModeChanged: (handler: (event: Event) => void) => void;
         attachUndoRedoStackModified: (handler: (event: Event) => void) => void;
         getCommandStack: () => Stack;
         getService: <T>(name: 'outline' | string) => Promise<T>;
         getSelection: () => ElementOverlay[];
         getDefaultPlugins: () => { contextMenu: ContextMenu };
+        getFlexSettings: () => { 
+            layer: string;
```
[the remark this answers is not in the document]
Comment (tobiasqueck) [reply packages/preview-middleware-client/types/sap.ui.rta.d.ts]: fixed with c3a8da0

File: packages/preview-middleware/src/base/flp.ts (hunk 2 above)
Comment (tobiasqueck) [reply packages/preview-middleware/src/base/flp.ts:185]: This is not coming from the yaml. The parameter options can only be set programmatically and is to be used to set scenario specific settings e.g. when initializing ADP, we set the `projectId`

File: packages/preview-middleware/test/unit/base/__snapshots__/flp.test.ts.snap (hunk 3 above)
Comment (tobiasqueck) [reply packages/preview-middleware/test/unit/base/__snapshots__/flp.test.ts.snap:288]: I know but do we really care? I can fix it, but I didn't see a reason to make the url nicer.

Comment (voicis) [reply packages/preview-middleware/test/unit/base/__snapshots__/flp.test.ts.snap:288]: It's not really visible so I suppose its ok.

**Review by voicis — APPROVED** [review]
- code changes looks good
- test coverage is good
- tested middleware locally

**Review by voicis — APPROVED** [review]

**Review by zdravko-georgiev — APPROVED** [review]
Code looks good
Very good test coverage
Manual testing ✅

**Pull request merged**